subsys/shell/backends/shell_mqtt.c

```diff
@@ -0,0 +1,824 @@
+	 * when internet is connected again
+	 */
+}
+
+/* Network connection event handler */
+static void network_evt_handler(struct net_mgmt_event_callback *cb, uint32_t mgmt_event,
+				struct net_if *iface)
+{
+	struct k_work_sync ws;
+
+	if (mgmt_event == NET_EVENT_L4_CONNECTED &&
+	    sh_mqtt->network_state == SHELL_MQTT_NETWORK_DISCONNECTED) {
+		LOG_WRN("Network %s", "connected");
+		sh_mqtt->network_state = SHELL_MQTT_NETWORK_CONNECTED;
+		(void)sh_mqtt_work_reschedule(&sh_mqtt->connect_dwork, K_SECONDS(1));
+	} else if (mgmt_event == NET_EVENT_L4_DISCONNECTED &&
+		   sh_mqtt->network_state == SHELL_MQTT_NETWORK_CONNECTED) {
+		(void)sh_mqtt_work_submit(&sh_mqtt->net_disconnected_work);
+	}
+}
+
+static void mqtt_evt_handler(struct mqtt_client *const client, const struct mqtt_evt *evt)
+{
+	switch (evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			sh_mqtt->transport_state = SHELL_MQTT_TRANSPORT_DISCONNECTED;
+			LOG_ERR("MQTT %s %d", "connect failed", evt->result);
+			break;
+		}
+
+		sh_mqtt->transport_state = SHELL_MQTT_TRANSPORT_CONNECTED;
+		LOG_WRN("MQTT %s", "client connected!");
+
+		break;
+	case MQTT_EVT_SUBACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT subscribe: %s", "error");
+			sh_mqtt->subscribe_state = SHELL_MQTT_NOT_SUBSCRIBED;
+			break;
+		}
+
+		LOG_WRN("MQTT subscribe: %s", "ok");
+		sh_mqtt->subscribe_state = SHELL_MQTT_SUBSCRIBED;
+		break;
+
+	case MQTT_EVT_UNSUBACK:
+		LOG_DBG("UNSUBACK packet id: %u", evt->param.suback.message_id);
+		sh_mqtt->subscribe_state = SHELL_MQTT_NOT_SUBSCRIBED;
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		LOG_WRN("MQTT disconnected: %d", evt->result);
+		sh_mqtt->transport_state = SHELL_MQTT_TRANSPORT_DISCONNECTED;
+		sh_mqtt->subscribe_state = SHELL_MQTT_NOT_SUBSCRIBED;
+		break;
+
+	case MQTT_EVT_PUBLISH: {
+		const struct mqtt_publish_param *pub = &evt->param.publish;
+		uint32_t size, payload_left, rb_free_space;
+
+		payload_left = pub->message.payload.len;
+		rb_free_space = ring_buf_space_get(&sh_mqtt->rx_rb);
+
+		LOG_DBG("MQTT publish received %d, %d bytes", evt->result, payload_left);
+		LOG_DBG("   id: %d, qos: %d", pub->message_id, pub->message.topic.qos);
+		LOG_DBG("   item: %s", log_strdup(pub->message.topic.topic.utf8));
+
+		/* For MQTT_QOS_0_AT_MOST_ONCE no acknowledgment needed */
+		if (pub->message.topic.qos == MQTT_QOS_1_AT_LEAST_ONCE) {
+			/* clang-format off */
+			struct mqtt_puback_param puback = {
+				.message_id = pub->message_id
+			};
+			/* clang-format on */
+
+			(void)mqtt_publish_qos1_ack(client, &puback);
+		}
+
+		while (payload_left > 0) {
+			/* Attempt to claim `payload_left` bytes of buffer in rb */
+			size = ring_buf_put_claim(&sh_mqtt->rx_rb, &sh_mqtt->rx_rb_ptr,
+						  payload_left);
+			/* Read `size` bytes of payload from mqtt */
+			size = mqtt_read_publish_payload_blocking(client, sh_mqtt->rx_rb_ptr, size);
+			/* Indicate that `size` bytes of payload has been written into rb */
```

Comment:
No error check here? In theory, the server could disconnect in the middle of payload sending which would end up in a virtually endless loop (as the payload_left would be increased by the error code value).

Reply:
I have added a check to flush the rb and return if negative value is returned.